### PR TITLE
Accessor and Mutator Methods for VAT Rates

### PR DESCRIPTION
**Pull request by AlexTheWizardL**
## Enhancements to CashCtrlLedger for Improved VAT Code Management

This pull request introduces significant enhancements to the `CashCtrlLedger` module, focusing on robust integration with the CashCtrl API for VAT rate management. We have added new functionalities for mirroring VAT rates onto CashCtrl, along with comprehensive error handling and code improvements to boost performance and maintainability.

- Encapsulated client management with getter and setter.
- Extended VAT code management functionalities (create, update, delete).
- Robust synchronization method to mirror VAT codes between local desired state and remote state.
- Comprehensive error handling and input validation.
- Detailed unit tests to cover a wide range of scenarios.

### Changes
- Implemented getter and setter for `client` to ensure type safety.
- Added `get_vat_codes`, `mirror_vat_codes`, `add_vat_code`, `update_vat_code`, and `delete_vat_code` methods.
- Each method is equipped with detailed docstrings and error handling.
- Added extensive tests using `pytest` to ensure functionality works as expected under various conditions.

- **Error Handling Enhancements**:
  - Enhanced error handling across methods to throw more specific exceptions, improving the clarity and response strategies in error scenarios.
  - Validation steps were added to setter methods to ensure that only valid instances of `CashCtrlClient` are assigned, preventing runtime errors.

- **Testing and Quality Assurance**:
  - Extensive test suite added in `tests/test_mirror_vat_code.py`, covering a variety of scenarios from basic functionality tests to complex integration and error handling tests.
  - Utilized `unittest.mock` for dependency injections, ensuring that unit tests are robust and not dependent on external systems.

### ChatGPT review:
<details> <summary>Click for ChatGPT prompt</summary>

### Overview
This review examines the newly added functions and tests in the `CashCtrlLedger` class which extends the `LedgerEngine`. These additions are intended to manage VAT codes through the CashCtrl system.

#### General Observations
- **Code Duplication**: The `client` instance is created twice in the constructor which is redundant. Consider initializing it once.
- **Error Handling**: It's good to see consistent error handling across the methods, especially with clear exception messages.
- **Docstrings**: The documentation is detailed and helpful, making the code easier to understand and maintain.

##### 1. Constructor Redundancy
- **Issue**: The client attribute is assigned twice redundantly.
- **Recommendation**: Initialize the _client attribute once and use the property setters/getters to manage it.

##### 2. Properties and Setters
- **Best Practice**: Good use of property decorators to manage encapsulation of internal state.
- **Issue**: The setter for client does not add much value as it merely checks the instance type. Consider whether this check could be enforced elsewhere to simplify the property.

##### 3. Method get_vat_codes
- **Best Practice**: Good transformation and renaming of dataframe columns to match expected output.
- **Potential Improvement**: Consider catching exceptions that might occur during API calls or data manipulation to enhance robustness.

##### 4. Method mirror_vat_codes
- **Complexity**: This method is quite complex and handles multiple functionalities (delete, update, create).
- **Recommendation**: Break down this method into smaller functions to improve readability and maintainability.

##### 5. Error Handling Consistency
- **Observation**: Different methods handle similar errors (like API failures) in slightly different ways.
- **Recommendation**: Standardize error handling across methods to maintain consistency and reduce code duplication.

##### 6. Tests
- **Coverage**: Tests cover a broad range of scenarios, which is excellent for ensuring robustness.
- **Improvement**: Some tests could be parameterized to reduce redundancy and increase the breadth of test cases with less code.

### Conclusion
These changes aim to align the CashCtrlLedger class with best practices observed in other successful open-source Python projects, improving both the readability and the robustness of the codebase.
</details>

### Review Request:
- Please review the error handling and suggest if specific exceptions are handled properly.
- Feedback on the test scenarios and any missing edge cases would be highly appreciated.
- @lasuk Please provide your insights and approval to merge this feature into the main branch.

### CODE REVIEW

File: cashctrl_ledger/ledger.py [hunk 1]
```diff
@@ -47,11 +205,128 @@ def add_price():
...
+    def update_vat_code(
+        self, code: str, rate: float, account: str,
+        inclusive: bool = True, text: str = "", name: str = "",
+    ):
+        """
+        Updates an existing VAT code in the CashCtrl account with new parameters. 
+        If a VAT code is already in use, some parameters like 'percentage' might not be changeable.
+
+        Parameters:
+            code (str): The unique identifier for the VAT code to be updated, max 50 characters.
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:276]: `code` is the name of the vat_code entry to modify. In pyledger, we do not know anything about CashCtrl's internal ids.

File: cashctrl_ledger/ledger.py [hunk 2]
```diff
@@ -47,11 +205,128 @@ def add_price():
...
+        try:
+            self.client.post("tax/create.json", data=payload)
+        except Exception as e:
+            raise Exception(
+                f"An error occurred while posting data to the server: {e}"
+            )
+        
+    def update_vat_code(
+        self, code: str, rate: float, account: str,
+        inclusive: bool = True, text: str = "", name: str = "",
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:269]: We do not need the argument 'name'.

File: cashctrl_ledger/ledger.py [hunk 3]
```diff
@@ -47,11 +205,128 @@ def add_price():
...
+            )
+
+        try:
+            self.client.post("tax/create.json", data=payload)
+        except Exception as e:
+            raise Exception(
+                f"An error occurred while posting data to the server: {e}"
+            )
+        
+    def update_vat_code(
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:267]: Most comments from `add_cat_code()` apply also to `update_vat_code()`.

File: cashctrl_ledger/ledger.py [hunk 4]
```diff
@@ -77,11 +352,34 @@ def delete_price():
         """
         raise NotImplementedError
 
-    def delete_vat_code():
-        """
-        Not implemented yet
+    def delete_vat_code(self, code: str):
         """
-        raise NotImplementedError
+        Deletes a VAT code from the tax system via a POST request to the API.
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:357]: Description should be
```
Deletes a VAT code from the remote CashCtrl account.
```

Instead of 
```
Deletes a VAT code from the tax system via a POST request to the API.
```

File: cashctrl_ledger/ledger.py [hunk 5]
```diff
@@ -77,11 +352,34 @@ def delete_price():
...
+        Exception
+            An exception is raised with a detailed error message if the API request fails.
+
+        Returns:
+        -------
+        None
+            This method does not return any value but will raise an exception if unsuccessful.
+        """
+        try:
+            self.client.post('tax/delete.json', {'ids': code})
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:379]: `code` maps to the string identifier (CashCtrl 'name'), we need to map it to the 'id'.

**Review by lasuk — COMMENTED** [review]
@AlexTheWizardL: Thanks for implementing the translation of pyledger 'vat_codes' to CashCtrl 'tax_rates'. 

The code is very clear and well written, looks professional, and is thoroughly tested.

Content-wise, we have not arrived at the finishing line yet. As discussed in our call this morning, there are several mapping errors: For example, we need to map pyledger 'account' numbers to CashCtrl 'accountId's, or pyledger string 'vat_codes' to CashCtrl integer tax code 'id'.

To expedite the review process and let you get going with modifications, I have limited my review to the getter and setter functions (get_vat_codes, add_vat_code, update_vat_code, remove_vat_codes). I have not yet review `mirror_vat_codes`, nor the unit tests.

As discussed this morning, I suggest to drop `mirror_vat_codes` from this merge request. Let's in a first step correct the getter and setter functions and their unit tests and complete their review. You can then add the `mirror_vat_code` function as a separate merge request.

File: cashctrl_ledger/ledger.py [hunk 6]
```diff
@@ -47,11 +82,120 @@ def add_price():
...
+            code (str): The VAT code to be added.
+            rate (float): The percentage rate of the VAT, must be between 0 and 1.
+            account (str): The account identifier to which the VAT is applied.
+            inclusive (bool): Determines whether the VAT is calculated as 'NET' 
+                            (True, default) or 'GROSS' (False).
+            text (str): Additional text or description associated with the VAT code.
+
+        Raises:
+            ValueError: If any of the inputs are invalid (wrong type or out of allowed range).
+            Exception: If there is an issue with the server connection or the API request.
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:102]: I've seen that you've removed the line 'Exception:  ....' in the docstring for `update_vat_code`. That makes sense to me. I suggest to remove it here as well.

File: cashctrl_ledger/ledger.py [hunk 7]
```diff
@@ -77,11 +221,21 @@ def delete_price():
...
-    def delete_vat_code():
+    def delete_vat_code(self, code: str):
         """
-        Not implemented yet
+        Deletes a VAT code from the remote CashCtrl account.
+
+        Parameters:
+        ----------
+        code : str
+            The VAT code name to be deleted.
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:231]: Please use the same documentation style for 'Parameters' as in other methods.

File: cashctrl_ledger/ledger.py [hunk 8]
```diff
@@ -77,11 +221,21 @@ def delete_price():
...
         """
-        raise NotImplementedError
+
+        remote_vats = self._client.list_tax_rates()
+        remote_vat_to_delete = remote_vats.loc[remote_vats['name'] == code]
+        delete_ids = ",".join(remote_vat_to_delete['id'].astype(str)) if not remote_vat_to_delete.empty else None
+
+        self._client.post('tax/delete.json', {'ids': delete_ids})
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:238]: Suggestion for improved readability:
- For if statements, use code blocks instead of one-line expressions
```
def delete_vat_code(self, code: str):
    remote_vats = self._client.list_tax_rates()
    to_delete = remote_vats.loc[remote_vats['name'] == code]
    if not to_delete.empty:
        delete_ids = ",".join(to_delete['id'].astype(str))
        self._client.post('tax/delete.json', {'ids': delete_ids})
```

File: cashctrl_ledger/ledger.py [hunk 9]
```diff
@@ -47,11 +82,120 @@ def add_price():
...
+        else:
+            account = None
+
+        payload = {
+            "name": code,
+            "percentage": rate*100,
+            "accountId": account,
+            "calcType": "NET" if inclusive else "GROSS",
+            "documentName": text,
+        }
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:118]: Suggestion to improve readability:
- Use a dict to map account 'number's to account 'id's
- TODO: align types between 'account' and account_map.keys().
```
def add_vat_code(...):
    accounts = self._client.list_accounts()
    account_map = accounts.set_index('number')['id'].to_dict()
    payload = {
        "name": code,
        "percentage": rate*100,
        "accountId": account_map.get(account, None),
        "calcType": "NET" if inclusive else "GROSS",
        "documentName": text,
    }
```

File: cashctrl_ledger/ledger.py [hunk 10]
```diff
@@ -47,11 +77,113 @@ def add_price():
...
+
+        if not isinstance(payload['documentName'], str) \
+                or len(payload['documentName']) > 50:
+            raise ValueError(
+                "Invalid text. It must be a string with a maximum of 50 characters."
+            )
+
+        self._client.post("tax/create.json", data=payload)
+        
+    def update_vat_code(
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:127]: Please align Parameter description and code in `update_vat_code` with the latest version of `add_vat_code`.

File: cashctrl_ledger/ledger.py [hunk 11]
```diff
@@ -77,11 +209,19 @@ def delete_price():
...
+        Parameters:
+            code (str): The VAT code name to be deleted.
         """
-        raise NotImplementedError
+        tax_rates = self._client.list_tax_rates()
+        to_delete = tax_rates.loc[tax_rates['name'] == code, 'id']
+
+        if len(to_delete) > 0:
+            delete_ids = ",".join(to_delete.astype(str))
+            self._client.post('tax/delete.json', {'ids': delete_ids})
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:224]: Let's add an else block, where we raise an exception "VAT code not found." if the given `code` does not match an existing one.

**Review by lasuk — COMMENTED** [review]
Thanks for implementing the changes. Here's two additional comments. We are almost there.

File: cashctrl_ledger/ledger.py [hunk 12]
```diff
@@ -47,11 +77,71 @@ def add_price():
...
+            text (str): Additional text or description associated with the VAT code.
+        """
+
+        accounts = self._client.list_accounts()
+        vat_account = accounts.loc[accounts['number'] == account]
+
+        if not vat_account.empty:
+            account = vat_account['id'].iloc[0]
+        else:
+            account = None
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:129]: Please use `account_map` similar to `add_vat_code()`.

File: cashctrl_ledger/ledger.py [hunk 13]
```diff
@@ -47,11 +77,71 @@ def add_price():
...
+
+        remote_vats = self._client.list_tax_rates()
+        remote_vat = remote_vats.loc[remote_vats['name'] == code]
+        remote_vat_id = remote_vat['id'].iloc[0] if not remote_vat.empty else None
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:133]: - Please ensure there is exactly one `remote_vat` entry before accessing item zero.
- Please drop the else part. The update method does not make any sense, if no remote_vat is found.

File: cashctrl_ledger/ledger.py [hunk 14]
```diff
@@ -47,11 +77,71 @@ def add_price():
...
+
+        payload = {
+            "id": remote_vat_id,
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:136]: More succinct to insert the value directly:
```
"id": remote_vat['id'].iloc[0]
```

File: cashctrl_ledger/ledger.py [hunk 15]
```diff
@@ -77,11 +167,21 @@ def delete_price():
...
         raise NotImplementedError
 
-    def delete_vat_code():
+    def delete_vat_code(self, code: str, allow_missing: bool = False):
         """
-        Not implemented yet
+        Deletes a VAT code from the remote CashCtrl account.
+
+        Parameters:
+            code (str): The VAT code name to be deleted.
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:175]: Parameter `allow_missing` is not documented.

**Review by lasuk — COMMENTED** [review]
Thanks for your changes. I've added a few new comments.

File: tests/test_vat_code.py [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+    updated_vat_codes = cashctrl_ledger.vat_codes()
+    updated_row = updated_vat_codes.loc[new_vat_code['code']]
+
+    assert all([
+        updated_row['text'] == new_vat_code['text'],
+        updated_row['account'] == new_vat_code['account'],
+        updated_row['rate'] == new_vat_code['rate'],
+        updated_row['inclusive'] == new_vat_code['inclusive']
+    ])
```
Comment (lasuk) [inline tests/test_vat_code.py:60]: Enhanced test function (same reasons as 'test_add_vat_code')
```
# Test updating a VAT code with valid inputs.
def test_update_vat_code():
    cashctrl_ledger = CashCtrlLedger()
    initial_vat_codes = cashctrl_ledger.vat_codes().reset_index()
    new_vat_code = {
        'code': "TestCode",
        'text': 'VAT 20%',
        'account': 2000,
        'rate': 0.20,
        'inclusive': True,
    }
    cashctrl_ledger.update_vat_code(**new_vat_code)
    updated_vat_codes = cashctrl_ledger.vat_codes().reset_index()
    # Diff between two DataFrames. See https://stackoverflow.com/questions/48647534
    modified_vat_codes = updated_vat_codes[
        ~updated_vat_codes.apply(tuple, 1).isin(initial_vat_codes.apply(tuple, 1))]
    assert len(modified_vat_codes) == 1, "Expected exactly one updated row"
    assert modified_vat_codes['id'].item() == new_vat_code['code']
    assert modified_vat_codes['text'].item() == new_vat_code['text']
    assert modified_vat_codes['account'].item() == new_vat_code['account']
    assert modified_vat_codes['rate'].item() == new_vat_code['rate']
    assert modified_vat_codes['inclusive'].item() == new_vat_code['inclusive']
```

File: tests/test_vat_code.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+def test_delete_vat_code():
+    """ Test deleting an existent VAT code. """
+    cashctrl_ledger = CashCtrlLedger()
+    cashctrl_ledger.delete_vat_code(code='TestCode')
+    updated_vat_codes = cashctrl_ledger.vat_codes()
+    deleted = 'TestCode' not in updated_vat_codes.index
+
+    assert deleted
```
Comment (lasuk) [inline tests/test_vat_code.py:69]: Write:
```
assert'TestCode' not in updated_vat_codes.index
```
instead of 

```
deleted = 'TestCode' not in updated_vat_codes.index

assert deleted
```

File: cashctrl_ledger/ledger.py [hunk 18]
```diff
@@ -77,11 +163,23 @@ def delete_price():
...
         """
-        raise NotImplementedError
+        tax_rates = self._client.list_tax_rates()
+        to_delete = tax_rates.loc[tax_rates['name'] == code, 'id']
+
+        if len(to_delete) > 0:
+            delete_ids = ",".join(to_delete.astype(str))
+            self._client.post('tax/delete.json', {'ids': delete_ids})
+        elif not allow_missing:
+            raise ValueError("VAT code not found.")
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:182]: Improved error message:
```
raise ValueError(f"VAT code '{code}' not found.")
```

File: cashctrl_ledger/ledger.py [hunk 19]
```diff
@@ -47,11 +77,67 @@ def add_price():
...
+            text (str): Additional text or description associated with the VAT code.
+        """
+
+        accounts = self._client.list_accounts()
+        account_map = accounts.set_index('number')['id'].to_dict()
+        remote_vats = self._client.list_tax_rates()
+        remote_vat = remote_vats.loc[remote_vats['name'] == code]
+
+        if len(remote_vat) != 1:
+            raise ValueError("There is no one or more than one VAT")
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:129]: Improve error message:
```
if len(remote_vat) < 1:
    raise ValueError(f"There is no VAT code '{code}'.")
elif len(remote_vat) > 1:
    raise ValueError(f"VAT code '{code}' is duplicated.")
```
(Maybe you find a better wording for 'duplicated')

File: tests/test_vat_code.py [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.delete_vat_code("TestCode")
+
+def test_update_non_existent_vat_raise_error():
+    """
+    Test updating a non existent VAT code should raise an error.
+    """
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(Exception):
```
Comment (lasuk) [inline tests/test_vat_code.py:84]: Better to narrow the test and check for the expected error class ('ValueError' instead of 'Exception')

File: tests/test_vat_code.py [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+        cashctrl_ledger.update_vat_code(code="TestCode", text='VAT 20%',
+            account=1, rate=0.02, inclusive=True
+        )
+
+def test_add_vat_with_not_valid_account_raise_error():
+    """
+    Test adding a VAT code with non existent account should raise an error.
+    """
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(Exception):
```
Comment (lasuk) [inline tests/test_vat_code.py:94]: Better to narrow the test and check for the expected error class ('RequestException' instead of 'Exception')

File: tests/test_vat_code.py [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+        cashctrl_ledger.add_vat_code(code="TestCode", text='VAT 20%',
```
Comment (lasuk) [inline tests/test_vat_code.py:95]: The error message is not very helpful in this case. I would expect something like "Account '7777' does not exist."

File: tests/test_vat_code.py (hunk 21 above)
Comment (lasuk) [reply tests/test_vat_code.py:94]: Applies also to cases further down.

File: tests/test_vat_code.py [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+    with pytest.raises(Exception):
+        cashctrl_ledger.add_vat_code()
+
+def test_update_vat_with_not_valid_input_raise_error():
+    """
+    Test updating a VAT code with not valid inputs should raise an error.
+    """
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(Exception):
+        cashctrl_ledger.update_vat_code()
```
Comment (lasuk) [inline tests/test_vat_code.py:132]: I am not convinced how this test adds anything. Calling any python function without specifying required arguments raises an error. I do not see anything specific to our setting.

**Review by lasuk — COMMENTED** [review]
Thanks. Back to you with additional comments.

File: cashctrl_ledger/ledger.py [hunk 24]
```diff
@@ -47,11 +77,75 @@ def add_price():
...
+        account_map = accounts.set_index('number')['id'].to_dict()
+        remote_vats = self._client.list_tax_rates()
+        remote_vat = remote_vats.loc[remote_vats['name'] == code]
+
+        if len(remote_vat) < 1:
+            raise ValueError(f"There is no VAT code '{code}'.")
+        elif len(remote_vat) > 1:
+            raise ValueError(f"VAT code '{code}' is duplicated.")
+        
+        if account_map.get(account, None) == None:
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:136]: This line is pretty un-pythonic :) 

Here's an improved version:
```python
    def update_vat_code(
        self, code: str, rate: float, account: str,
        inclusive: bool = True, text: str = ""
    ):
        """
        Updates an existing VAT code in the CashCtrl account with new parameters.

        Parameters:
            code (str): The VAT code to be updated.
            rate (float): The VAT rate, must be between 0 and 1.
            account (str): The account identifier to which the VAT is applied.
            inclusive (bool): Determines whether the VAT is calculated as 'NET'
                            (True, default) or 'GROSS' (False).
            text (str): Additional text or description associated with the VAT code.
        """
        # Find remote account id
        accounts = self._client.list_accounts()
        account_map = accounts.set_index('number')['id'].to_dict()
        if account not in account_map:
            raise ValueError(f"Account '{account}' does not exist.")

        # Find remote tax id
        remote_vats = self._client.list_tax_rates()
        remote_vat = remote_vats.loc[remote_vats['name'] == code]
        if len(remote_vat) < 1:
            raise ValueError(f"There is no VAT code '{code}'.")
        elif len(remote_vat) > 1:
            raise ValueError(f"VAT code '{code}' is duplicated.")

        # Update remote tax record
        payload = {
            "id": remote_vat['id'].item(),
            "percentage": rate*100,
            "accountId": account_map[account],
            "calcType": "NET" if inclusive else "GROSS",
            "name": code,
            "documentName": text,
        }
        self._client.post("tax/update.json", data=payload)
```

File: cashctrl_ledger/ledger.py [hunk 25]
```diff
@@ -47,11 +77,75 @@ def add_price():
...
+        payload = {
+            "name": code,
+            "percentage": rate*100,
+            "accountId": account_map.get(account, None),
+            "calcType": "NET" if inclusive else "GROSS",
+            "documentName": text,
+        }
+
+        if payload['accountId'] == None:
+            raise ValueError(f"Account '{account}' does not exist.")
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:106]: First assigning 'accountId' with default None, and later checking whether the value is 'None' is a bit twisted.

Here's a suggestion for better readability:
```python
    def add_vat_code(
        self, code: str, rate: float, account: str, inclusive: bool = True,
        text: str = ""
    ):
        """
        Adds a new VAT code to the CashCtrl account.

        Parameters:
            code (str): The VAT code to be added.
            rate (float): The VAT rate, must be between 0 and 1.
            account (str): The account identifier to which the VAT is applied.
            inclusive (bool): Determines whether the VAT is calculated as 'NET'
                            (True, default) or 'GROSS' (False).
            text (str): Additional text or description associated with the VAT code.
        """
        accounts = self._client.list_accounts()
        account_map = accounts.set_index('number')['id'].to_dict()
        if account not in account_map:
            raise ValueError(f"Account '{account}' does not exist.")
        payload = {
            "name": code,
            "percentage": rate*100,
            "accountId": account_map[account],
            "calcType": "NET" if inclusive else "GROSS",
            "documentName": text,
        }
        self._client.post("tax/create.json", data=payload)
```

File: tests/test_vat_code.py [hunk 26]
```diff
@@ -0,0 +1,118 @@
+"""
+Unit tests for vat codes operation methods.
```
Comment (lasuk) [inline tests/test_vat_code.py:2]: Slightly improved unit docstring:
```
Unit tests for vat codes accessor and mutator methods.
```

File: cashctrl_ledger/ledger.py [hunk 27]
```diff
@@ -77,11 +171,23 @@ def delete_price():
...
         """
-        raise NotImplementedError
+        tax_rates = self._client.list_tax_rates()
+        to_delete = tax_rates.loc[tax_rates['name'] == code, 'id']
+
+        if len(to_delete) > 0:
+            delete_ids = ",".join(to_delete.astype(str))
+            self._client.post('tax/delete.json', {'ids': delete_ids})
+        elif not allow_missing:
+            raise ValueError(f"VAT code {code} not found.")
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:190]: Let's use the same error message as on L132, I find it more explicit:
```
raise ValueError(f"There is no VAT code '{code}'.")
```

File: tests/test_vat_code.py [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.delete_vat_code("TestCode")
+
+# Test updating a non existent VAT code should raise an error.
+def test_update_non_existent_vat_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.update_vat_code(code="TestCode", text='VAT 20%',
+            account=1, rate=0.02, inclusive=True
```
Comment (lasuk) [inline tests/test_vat_code.py:79]: Please use account 2200 (account 1 does usually not exist, it was introduced in your earlier tests).

File: tests/test_vat_code.py [hunk 29]
```diff
@@ -0,0 +1,118 @@
...
+    with pytest.raises(ValueError):
+        cashctrl_ledger.update_vat_code(code="TestCode", text='VAT 20%',
+            account=7777, rate=0.02, inclusive=True
+        )
+
+# Test deleting a VAT code with not valid inputs should raise an error.
+def test_delete_vat_with_not_valid_input_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.delete_vat_code(code="")
```
Comment (lasuk) [inline tests/test_vat_code.py:102]: Drop as redundant (tests the same functionality as L74:80).

File: tests/test_vat_code.py [hunk 30]
```diff
@@ -0,0 +1,118 @@
...
+            account=7777, rate='kkk', inclusive=3
+        )
+
+# Test updating a VAT code with not valid inputs should raise an error.
+def test_update_vat_with_not_valid_input_raise_error():
+    cashctrl_ledger = CashCtrlLedger()
+    with pytest.raises(ValueError):
+        cashctrl_ledger.update_vat_code(code="", text='',
+            account=7777, rate='kkk', inclusive=3
+        )
```
Comment (lasuk) [inline tests/test_vat_code.py:118]: Please drop the last two tests, I do not see how they add value.

**Review by lasuk — COMMENTED** [review]
Thanks. Back to you with a few final, minor comments.

**Comment by lasuk** [comment]
Merging. Thanks a lot for the contribution, endurance and great result!
